text/3621-derive-smart-pointer.md

```diff
@@ -0,0 +1,772 @@
+`DispatchFromDyn`.
+
+[`Cell`]: https://doc.rust-lang.org/stable/core/cell/struct.Cell.html
+
+## What about `#[pointee]`?
+
+This RFC currently proposes to mark the generic parameter used for dynamic
+dispatch with `#[pointee]`. For convenience, the RFC proposes that this is only
+needed when there are multiple generic parameters.
+
+There are potential use-cases for smart pointers with additional generic
+parameters. Specifically, the `ListArc` type used by the linked lists currently
+has an additional const generic parameter to allow you to use the same
+refcounted value with multiple lists. People have argued that it would be
+better to change this to a generic type instead of a const generic, so it would
+be useful to keep the option of having multiple generic types on the struct.
+
+### Conflicts with third-party derive macros
+
+The `#[pointee]` attribute could in principle conflict with other derive macros
+that also wish to annotate one of the parameters with an attribute called
+`#[pointee]`. To disambiguate such cases, we also allow the attribute to be
+spelled `#[smart_pointer::pointee]`.
+
+It is an error to specify both `#[pointee]` and `#[smart_pointer::pointee]`, so
+both macros must support this kind of disambiguation.
+
+Another way to avoid conflicts between `#[derive(SmartPointer)]` and third-party
+macros is to always assume that the first generic parameter is the pointee.
+This RFC does not propose that solution because:
+
+* It prevents the pointee from having a default unless it is the only parameter,
+  because parameters with a default must come last.
+* If logic such as "the first parameter" becomes commonplace in macro design,
+  then it does not really solve the issue with conflicts: you could have two
+  macros that both assume that the first parameter is special. And this kind of
+  conflict will be more common than attribute conflicts, because the attribute
+  will only conflict if both macros use an attribute of the same name.
+
+The authors are not aware of any macros using a `#[pointee]` attribute today.
+
+## Derive macro or not?
+
+Stabilizing this as a derive macro more or less locks us in with the decision
+that the compiler will use traits to specify which types are compatible with
+trait objects. However, one could imagine other mechanisms. For example, stable
+Rust currently has logic saying that any struct where the last field is `?Sized`
+will work with unsizing operations. (E.g., if `Wrapper` is such a struct, then
+you can convert from `Box<Wrapper<[u8; 10]>>` to `Box<Wrapper<[u8]>>`.) That
+mechanism is not specified using a trait.
+
+However, using traits for this functionality seems to be the most flexible. To
+solve the unresolved questions, we most likely need to constrain the
+implementations of these traits for `Pin` with stricter trait bounds than what
+is specified on the struct. That will get much more complicated if we use a
+mechanism other than traits to specify this logic.
```

Comment:
Continuing https://github.com/rust-lang/rfcs/pull/3621#issuecomment-2096626434 by @clarfonthey 

I'm not convinced about `#[repr(smart_pointer)]`. It's pretty weird for a `repr` to implement traits on the type. And like I explained in this new section, I think it is unlikely that we will not end up using traits for this functionality.

I think we can get pretty good errors from the derive macro. The `DispatchFromDyn` type already has special built-in checks in the compiler, and we can just adjust the errors that it emits to get reasonable errors.

Other than that, did you see [commit f7758256](https://github.com/rust-lang/rfcs/pull/3621/commits/f77582569b942595758311d2232aab3f0d8f3bd6)? It's not correct that only smart pointers implement both traits.

Reply:
Note that them existing as traits is mostly just a quirk of the existing implementation. Since they're unstable, the traits might as well not exist.

Reply:
Of course, but I imagine that `#[derive(SmartPointer)]` will not be the end of evolution of this feature. At some point we will hopefully stabilize the underlying traits as well, and I think it is quite likely that we still want them to be traits when we do so.

Reply:
Honestly, I'm not quite sure how that would be useful, but I'll ask anyway: what kind of generic code would you expect to be able to use these traits for?

Reply:
It's not so much that I think it would be useful for generic code. Rather, I think it is useful to use traits because it gives us a lot of control over _when_ the special traits are implemented.

Imagine we go for the `#[repr(smart_pointer)]` solution and also have a `#[repr(transparent_container)]` for the case of `Cell`. Then each type has essentially three possible choices:

1. No `repr` annotation. The type is never usable with dynamic dispatch.
2. The `#[repr(smart_pointer)]` annotation. The type is always usable with dynamic dispatch.
3. The `#[repr(transparent_container)]` annotation. The type is usable with dynamic dispatch whenever its contents are.

However, these choices are not sufficiently flexible due to the `Pin` issue. If we go with [the `StableDeref` solution](https://github.com/Darksonn/rfcs/blob/derive-smart-pointer/text/3621-derive-smart-pointer.md#stablederef), which is the most elegant solution, then we need a fourth option of "the type is usable with dynamic dispatch whenever the inner smart pointer implements `StableDeref`".

Using traits, this is no issue. You just add `T: StableDeref` as a trait bound to the implementation and that's it.

Reply:
Ah, I see. From that perspective, you're definitely right that having traits would be helpful long-term, even though it's still mostly a compiler detail. I think this analysis would be good to add to that section of the RFC; I think that deriving would make sense here, but that then leads back to the original issue about the combined deriving.

Of course, I guess that derive macros could ultimately be deprecated anyway if we decide to split them up later. Although it's unprecedented for the standard library, `proc_macro`s can do whatever they want, so, we already see individual crates doing stuff like this.

Reply:
I did try to explain this in the section I added yesterday, but I will elaborate in more details.

Reply:
(I also hadn't read the section before you mentioned it, so, I'll double check I didn't just miss the details there.)